WK2-261: documentation guides improvements after feedback (#14)
wealthkernel/api-documentation-guides: docs/webhooks/Secrets.md

We sign our webhooks using a secret, to prove that it is WealthKernel sending you requests, and not a third party. You can generate this secret through our portal. The secret will only be shown to you once, and is not retrievable, so make sure you record it at this point. The secret should be stored securely.

The secret is the same for all subscriptions and event types, and consists of a series of cryptographically strong bytes encoded into base64. It is not possible to receive webhooks without first generating a secret.

Secrets can be disabled, should the need arise, for example if your secret becomes compromised. If you disable your only secret, all webhook traffic will stop until a new secret is generated.

## Secret expiration

Secrets expire after 3 months, however you may have up to two active secrets concurrently, allowing you to continue to receive webhooks without any downtime. The expiration time of your secret can be viewed through our portal at any time.

> The webhook signature header will contain one or more HMAC (hash-based message authentication code) values, depending on the number of active secrets at a given time. If you have two active secrets and one of them expires, from that point you will only receive one HMAC in the signature. If all secrets expire you will stop receiving webhooks.

Reviewer: ```suggestion
The webhook signature header will contain one or more HMAC (hash-based message authentication code) values, depending on the number of active secrets at a given time.`If you have two active secrets and one of them expires, from that point you will only receive HMAC(s) for the active secret.` If all secrets expire you will stop receiving webhooks.
```

Author: @tom-meader-wk changed

We sign our webhooks using a secret, to prove that it is WealthKernel sending you requests, and not a third party. You can generate this secret through our portal. The secret will only be shown to you once, and is not retrievable, so make sure you record it at this point. The secret should be stored securely.

The secret is the same for all subscriptions and event types, and consists of a series of cryptographically strong bytes encoded into base64. It is not possible to receive webhooks without first generating a secret.

Secrets can be disabled, should the need arise, for example if your secret becomes compromised. If you disable your only secret, all webhook traffic will stop until a new secret is generated.

## Secret expiration

Secrets expire after 3 months, however you may have up to two active secrets concurrently, allowing you to continue to receive webhooks without any downtime. The expiration time of your secret can be viewed through our portal at any time.

> The webhook signature header will contain one or two HMAC (hash-based message authentication code) values, depending on the number of active secrets at a given time. If you have two active secrets and one of them expires, from that point you will only receive one HMAC in the signature. If all secrets expire you will stop receiving webhooks.

Reviewer 2: There could be more than two HMACs as we support multiple schemes.

We recommend to have one active secret and generate a new one when the expiration date of the secret gets close.

## Using a secret to verify a webhook

On every webhook request there is a header called `Webhook-Signature`, which is composed of:
1. A timestamp in [UNIX seconds](https://en.wikipedia.org/wiki/Unix_time) prefixed by `t=`.
1. One HMAC prefixed by `v1=`.

> 1. An additional HMAC prefixed by `v1=` if there are two active secrets.

Reviewer 2: There could be more than two HMACs as we support multiple schemes.

Reviewer 2: Could we use bullet points here instead of a numbered list e.g.

- A timestamp in [UNIX seconds](https://en.wikipedia.org/wiki/Unix_time) prefixed by `t=`.
- HMAC foreach secret and scheme prefixed by `${scheme}=`. e.g. `v1=`.

> The value is the same for all subscriptions and event types, and consists of a series of cryptographically strong bytes encoded into base64. It is not possible to receive webhooks without first generating a secret.

Reviewer 3: I find it clearer to have `The secret` instead of `The value`.  I didn't realise it was referencing the secret value at first sight

Secrets can be disabled, should the need arise, for example if your secret becomes compromised. If you disable your only secret, all webhook traffic will stop until a new secret is generated.

## Secret expiration

Secrets expire after 3 months, however you may have up to two active secrets concurrently, allowing you to continue to receive webhooks without any downtime. The expiration time of your secret can be viewed through our portal at any time.

> The webhook signature will contain one or two HMAC values, depending on the number of active secrets at a given time. If you have two active secrets and one of them expires, from that point you will only receive one HMAC in the signature. If all secrets expire you will stop receiving webhooks.

Reviewer 3: Maybe mention that the signature will be sent in the webhook's header

We recommend to have one active secret and generate a new one when the expiration date of the secret gets close.

## Using a secret to verify a webhook

On every webhook request there is a header called `Webhook-Signature`, which is composed of:
1. A timestamp in [UNIX seconds](https://en.wikipedia.org/wiki/Unix_time) prefixed by `t=`.

> 1. At least one HMAC (hash-based message authentication code) prefixed by `v1=`.

Reviewer 3: I think the HMAC meaning should appear in the first place we use the acronym.

> 

Reviewer 3: How about three points? 1. Timestamp, 2. HMAC / signature 3. Additional HMAC / signature if there are two active secrets

1. Get the hexadecimal encoding of the genereated HMAC.
1. Compare the hexadecimal encoded HMAC(s) from the first step to the hexadecimal encoded HMAC(s) you have generated. Check if the timestamp is within your tolerance (to protect from [replay attacks](https://en.wikipedia.org/wiki/Replay_attack))

Note that if you have multiple secrets you might need to compare more than one HMAC.

<!-- theme: info -->
> A constant time string comparison here is recommended to avoid [timing attacks.](https://en.wikipedia.org/wiki/Timing_attack)

Here's the pseudocode to verify a webhook signature with one HMAC:
```
signature = getRequestHeader("Webhook-Signature")
body = getRequestBody()
secretBase64 = getSecret()
secretDecoded = decodeBase64(secretBase64)

signatureSplitByComma = signature.Split(",")
timestamp = signatureSplitByComma[0].Split("t=")[1]
signatureHmac = signatureSplitByComma[1].Split("v1=")[1]

content = concatenate(body, timestamp)
contentHmac = getHmacSha256(content)
contentHmacHexadecimal = toHexadecimal(contentHmac)

> isVerified = equals(signatureHmac, contentSignedHexadecimal)

Reviewer 3: Should `contentSignedHexadecimal` be renamed to `contentHmacHexadecimal`?
